src/Java.Interop.NamingCustomAttributes/Java.Interop/IJniNameProviderAttribute.cs

```diff
@@ -0,0 +1,8 @@
+﻿using System;
+namespace Java.Interop
+{
+	public interface IJniNameProviderAttribute
```

Comment:
Should this be `public`? I'm not sure that it *does*, as [the `.projitems` is directly included into `Mono.Android.dll`](https://github.com/xamarin/xamarin-android/blob/2b4039a8d67e94b453889a148d22bb006f8835a4/src/Mono.Android/Mono.Android.csproj#L84). As such, they're in the same assembly.

Reply:
Naming: an `Attribute` suffix implies that the type is a custom attribute, which isn't the case here. I would thus suggest the interface name `IJniNameProvider`...which *still* isn't right, as that implies that `IJniNameProvider.Name` would be a JNI name, which *isn't the case*: a JNI name would be e.g. `java/lang/String` or `Ljava/lang/String;`, where our code expects `java.lang.String`...though would work with `java/lang/String`...

I'm thus torn.

However, given that our code should work sanely with `java/lang/String`, the name `IJniNameProvider` should suffice. Alternatively, perhaps `IRegisterAttributeNameProvider` (though I prefer `IJniNameProvider`).

Reply:
I have never heard of such implication. Can you provide examples for "IxxxxAttribute" is an attribute class?

Reply:
The [Framework Design Guidelines section on Attributes](https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/attributes) says:

> ✓ **DO** name custom attribute classes with the suffix "Attribute." +

Meanwhile, convention is that interfaces start with an `I`. There is thus "tension" in naming: does the leading `I` mean it's an interface, or does the `Attribute` suffix mean it's an attribute for use on interface?

I don't know.

So if I look at mono:

```bash
$ git grep 'class\s*I\w*Attribute\>'
```

there is an interesting hit on a custom attribute beginning with `I`:

* [`IDispatchImplAttribute`](https://github.com/mono/mono/blob/67c9ed311c0e214f9c58fdc96ad6264927dd6243/mcs/class/referencesource/mscorlib/system/runtime/interopservices/attributes.cs#L233)

Meanwhile, there are *also* interesting hits on *interfaces* ending in `Attribute`:

```bash
$ git grep 'interface\s*I\w*Attribute\>'
```

Which hits:

* [`IContractBehaviorAttribute.cs `](https://github.com/mono/mono/blob/67c9ed311c0e214f9c58fdc96ad6264927dd6243/mcs/class/System.ServiceModel/System.ServiceModel.Description/IContractBehaviorAttribute.cs#L35)
* [`IContextAttribute`](https://github.com/mono/mono/blob/67c9ed311c0e214f9c58fdc96ad6264927dd6243/mcs/class/corlib/System.Runtime.Remoting.Contexts/IContextAttribute.cs#L38)
* [`IContractBehaviorAttribute`](https://github.com/mono/referencesource/blob/e458f8df6ded689323d4bd1a2a725ad32668aaec/System.ServiceModel/System/ServiceModel/Description/IContractBehaviorAttribute.cs#L9)

What this means is that if you see `IFooAttribute`, you can't assume you know what it is: it could be *either* a custom attribute, *or* an interface. Neither convention "wins".

Reply:
IDispatchImplAttribute should not start name with I because "DO use I for interfaces".

... You'd notice that your logic is broken: it says "DO name custom attribute classes with the suffix Attribute." but it never says "DO NOT USE Attribute for anything else than Attributes".

IDispatchImplAttribute can retain that name because there is no guideline that says "DO NOT USE I for non-interfaces".

Though, I'm in general okay with IJniNameProvider, but it feels awkward and looks mismatching as compared to what WCF did in the past. The uses of this interface explicitly targets custom attributes. If you still have reason to prefer IJniNameProvider, I'm all ears.

Visibility wise, I cannot think of any reason to avoid public interface. I don't think having internal interfaces is good. There is nothing that developers expect ABI on these types. We could just stop using the interface IF there is problem, and Name properties are always there.